crates/chia-client/src/request_map.rs

```diff
@@ -0,0 +1,65 @@
+use std::{collections::HashMap, sync::Arc};
+
+use chia_protocol::Message;
+use tokio::sync::{oneshot, Mutex, OwnedSemaphorePermit, Semaphore};
+
+#[derive(Debug)]
+pub struct Request {
+    sender: oneshot::Sender<Message>,
+    _permit: OwnedSemaphorePermit,
+}
+
+impl Request {
+    pub fn send(self, message: Message) {
+        self.sender.send(message).ok();
+    }
+}
+
+#[derive(Debug)]
+pub struct RequestMap {
+    items: Mutex<HashMap<u16, Request>>,
+    semaphore: Arc<Semaphore>,
+}
+
+impl RequestMap {
+    pub fn new() -> Self {
+        Self {
+            items: Mutex::new(HashMap::new()),
+            semaphore: Arc::new(Semaphore::new(u16::MAX as usize)),
+        }
+    }
+
+    pub async fn insert(&self, sender: oneshot::Sender<Message>) -> u16 {
+        let permit = self
+            .semaphore
+            .clone()
+            .acquire_owned()
+            .await
+            .expect("semaphore closed");
```

Comment:
I imagine the semaphore is use to ensure no more than 65k coroutines enter this function simultaneously. Although, the mutex lock below would serialize them all anyway, it ensures there's always a free slot in the hash map.

This is subtle enough to warrant a comment I think